src/Opw.PineBlog.EntityFrameworkCore/BlogEntityDbContext.cs

```diff
@@ -23,5 +31,162 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            Posts.Add(post);
+            return SaveChangesAsync(cancellationToken);
+        }
+
+        public Task<Author> GetAuthorByUsernameAsync(string username)
+        {
+            return Authors.SingleOrDefaultAsync(a => a.UserName.Equals(username));
+        }
+
+        public Task<Result<int>> DeletePostAsync(Post post, CancellationToken cancellationToken)
+        {
+            Posts.Remove(post);
+            return SaveChangesAsync(true, cancellationToken);
+        }
+
+        public async Task<IList<Post>> GetPostListAsync(bool includeUnpublished, Pager pager, string category, PineBlogOptions options, CancellationToken cancellationToken)
+        {
+            var pagingUrlPartFormat = options.PagingUrlPartFormat;
+            var predicates = new List<Expression<Func<Post, bool>>>();
+            if (!includeUnpublished)
+                predicates.Add(p => p.Published != null);
+            if (!string.IsNullOrWhiteSpace(category))
+            {
+                predicates.Add(p => p.Categories.Contains(category));
+                pagingUrlPartFormat += "&" + string.Format(options.CategoryUrlPartFormat, category);
+            }
+            var posts = await GetPagedListAsync(predicates, pager, pagingUrlPartFormat, cancellationToken);
+            return posts;
+        }
+
+        private async Task<IList<Post>> GetPagedListAsync(IList<Expression<Func<Post, bool>>> predicates, Pager pager, string pagingUrlPartFormat, CancellationToken cancellationToken)
+        {
+            var skip = (pager.CurrentPage - 1) * pager.ItemsPerPage;
+
+            var countQuery = Posts.Where(_ => true);
+            foreach (var predicate in predicates)
+                countQuery = countQuery.Where(predicate);
+
+            var count = await countQuery.CountAsync(cancellationToken);
+
+            pager.Configure(count, pagingUrlPartFormat);
+
+            var query = Posts.Where(_ => true);
+
+            foreach (var predicate in predicates)
+                query = query.Where(predicate);
+
+            query = query.Include(p => p.Author)
+                .OrderByDescending(p => p.Published)
+                .Skip(skip)
+                .Take(pager.ItemsPerPage);
+
+            return await query.ToListAsync(cancellationToken);
+        }
+
+        public Task<Post> GetPostByIdAsync(Guid id, CancellationToken cancellationToken)
+        {
+            return Posts.SingleOrDefaultAsync(e => e.Id.Equals(id), cancellationToken);
+        }
+
+        public Task<Post> GetPostBySlugAsync(string slug, CancellationToken cancellationToken)
+        {
+            return Posts
+                .Include(p => p.Author)
+                .Where(p => p.Published != null)
+                .Where(p => p.Slug.Equals(slug))
+                .SingleOrDefaultAsync(cancellationToken);
+        }
+
+        public async Task<Result<Post>> PublishPostAsync(Guid id, CancellationToken cancellationToken)
```

Comment:
`Publish` is business logic and should be handled in the `Command`

Reply:
Agreed. `PublishPostCommand` will implement this by `UpdatePostAsync`.